md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2357 @@
+   that instruction are required to have been pushed within the same region,
+   including regions nested inside it.
+
+The requirements, plus the various minor requirements, can be validated in a
+single linear pass as follows:
+
+The following data structures are created:
+ - A *type stack*, with entries each containing a [type], for tracking the
+   sequence of values that will be on the value stack at execution time.
+ - A *control-flow stack*, with entries each containing
+    - A *limit* integer value, which is an index into the type stack indicating
+      the boundary between values pushed before the current region and values
+      pushed inside the current region.
+    - An optional [type] sequence, which is used to check that all exits from
+      the current region leave the same sequence of types on the type stack.
+
+The following invariants are required to be preserved:
+ - Whenever an element of either stack is to be popped, that stack is required
+   to be non-empty.
+ - Whenever an element of either stack is to be accessed by index, the index is
+   required to be within the bounds of that stack.
+ - When a [type] is to be popped from the type stack, the length of the type
+   stack is required to be greater than the control-flow stack top's limit
+   value.
+
+The type stack begins empty. The control-flow stack begins with one entry, with
+the [type] sequence consisting of the return types of the function, and with the
+limit value being zero.
+
+The function's instruction sequence is required to be non-empty, and the last
+instruction in the sequence is required to be an [`end`](#end).
+
+For each instruction in the body, in sequence order:
+ - For each construct in the instruction's signature's operands in reverse
+   order:
+    - If the element is a [type], a type is popped from the type stack and
+      required to be the same.
+    - If the element is a list of type parameters, and the parameters are not
+      yet bound, a type for each type parameter in the list in reverse order is
+      popped from the type stack and bound to it.
+    - Otherwise, a type for each type in the list in reverse order is popped
+      from the type stack and required to be the same.
+   The popped types in reverse (again) order form the *operand sequence*.
+ - Unless the instruction is a [control-flow barrier][Q], then for each
+   construct in the instruction's signature's returns:
+    - If the element is a [type], a type is pushed onto the type stack.
+    - If the element is a list of type parameters, they are required to be
+      bound, and the type bound to each type parameter is pushed onto the type
+      stack.
+ - If the instruction has a **Validation** clause, its requirements are
+   required.
+ - If the instruction's signature has no return types, or it is a
+   [control-flow barrier][Q], the new length of the type stack is required to be
+   at most the control-flow stack top's limit value.
+
```

Comment:
The problem is not just one of accessing these values, rather this constraint does not even permit values to be stored in locals for further use, even if they could be accessed. Even with a `get_value` operator, e.g. `call $fn_returning_two_values, get_value 1, store_local $l1` would be invalid.

A `drop_all` operator seems possible for the case in which none of the results were required. A `multiple_set_local` operator seems possible that would transfer all the values into locals but then the call operator might as well just write all the results directly into locals and we end up at the expressionless encoding.

It is possible to convert an AST with multiple value support into a stack machine code but perhaps what is missing are practical constraints that would limit this stack machine code to being presentable as an AST. That is, perhaps wasm can be an AST or it can be a stack machine but not both, and this has been heading down the path of the stack machine. Perhaps there are some constrains, but if they are not known to anyone then that seem a blocker, unless people want to concede the AST text format and have a stack machine text format.